[virt-api, migration-create-admitter]: Use KubeVirt generated clientset

Acedus opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
**Before this PR:**

MigrationCreateAdmitter:
1. Had exported fields and could be instantiated incorrectly.
2. Used the old `kubecli.KubevirtClient` clientset.
3. Used gomock for unit tests.
4. Used `v1 api` to create base VMIs for tests.

**After this PR:**
MigrationCreateAdmitter:
1. Has a constructor function and only unexported fields.
2. Uses KubeVirt's generated clientset.
3. Uses `libvmi` to create base VMIs for tests.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

This is a follow up to https://github.com/kubevirt/kubevirt/pull/11897.

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
6. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

Acedus commented:
/sig code-quality

Acedus commented:
/cc @orelmisan

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/4f66cabeeb80c2d4daa05108d431c194c17477c2..717f608dedfde3673f789cbdc15b7e3402f27cf8) - Separate commit for moving the test to its own separate package.

orelmisan commented on pkg/virt-api/webhooks/validating-webhook/admitters/migration-create-admitter_test.go:
```diff
@@ -34,91 +33,65 @@ import (
-	// Mock VirtualMachineInstanceMigration
-	var ctrl *gomock.Controller
-	var virtClient *kubecli.MockKubevirtClient
+	var virtClient *kubevirtfake.Clientset
```
Please consider dropping the shared variable and use a client per test.
It will enable dropping the `BeforeEach` function and allow a different approach for initializing the client:
```go
virtClient := kubevirtfake.NewSimpleClientset(<objects you want to have in the cluster>)
```

For example, vmi's could already be present in the fake client, without calling their matching "Create()" method.

https://github.com/kubevirt/kubevirt/blob/05bb90e20d0a9064ff27309d67e1828030266963/pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter_test.go#L116

Acedus replied:
Good idea, allow me to offer a slightly different approach.

How about keeping everything in `BeforeEach` and call `virtClient.KubevirtV1().VirtualMachineInstances(vmi.Namespace).UpdateStatus(context.Background(), vmi, metav1.UpdateOptions{})` when necessary (e.g. after Status updates)?

Acedus replied:
Discussed offline, it would in fact be cleaner without `BeforeEach`.

orelmisan commented on pkg/virt-api/webhooks/validating-webhook/admitters/migration-create-admitter_test.go:
```diff
@@ -34,91 +33,65 @@ import (
 		inFlightMigration := v1.VirtualMachineInstanceMigration{
 			ObjectMeta: metav1.ObjectMeta{
 				Namespace: vmi.Namespace,
+				Labels:    map[string]string{v1.MigrationSelectorLabel: vmi.Name},
```
Could you please add reasoning for this addition? (IIRC it is the result of the migration mutator)

Acedus replied:
Fakeclient doesn't pass it through mutation so it never hits `addMigrationSelectorLabel()`, I could add a prepend reactor but that would be an overkill IMHO so I'll be keeping it like so otherwise the test can't run properly.

orelmisan replied:
I meant adding an explanation to the commit message why is this change needed.

orelmisan commented on pkg/virt-api/webhooks/validating-webhook/admitters/migration-create-admitter_test.go:
```diff
@@ -289,3 +230,21 @@ var _ = Describe("Validating MigrationCreate Admitter", func() {
+		},
+	}
+
+	return ar, err
```
nit: Since this is the happy path:
```suggestion
	return ar, nil
```

Acedus replied:
Done.

orelmisan reviewed (commented):
Thank you for the PR @Acedus.
Please see the inline comments.

Acedus commented on pkg/virt-api/webhooks/validating-webhook/admitters/migration-create-admitter.go:
```diff
@@ -39,7 +39,13 @@ import (
 )
 
 type MigrationCreateAdmitter struct {
-	VirtClient kubecli.KubevirtClient
```
Done

orelmisan replied:
I've received a comment to replace `factory function` to `constructor` in https://github.com/kubevirt/kubevirt/pull/12654#issuecomment-2310661208.

orelmisan commented on pkg/virt-api/webhooks/validating-webhook/admitters/migration-create-admitter_test.go:
```diff
@@ -289,3 +230,21 @@ var _ = Describe("Validating MigrationCreate Admitter", func() {
 	})
 })
+
+func newAdmissionReviewForVMIMCreate(migration *v1.VirtualMachineInstanceMigration) (*admissionv1.AdmissionReview, error) {
```
Please consider introducing this change in a dedicated commit.

Acedus replied:
Done.

orelmisan commented on pkg/virt-api/webhooks/validating-webhook/admitters/migration-create-admitter_test.go:
```diff
@@ -30,30 +30,18 @@ import (
 	v1 "kubevirt.io/api/core/v1"
 	kubevirtfake "kubevirt.io/client-go/generated/kubevirt/clientset/versioned/fake"
 
+	"kubevirt.io/kubevirt/pkg/libvmi"
```
Please consider squashing this commit with the one introducing the fake client.

Acedus replied:
Done.

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/7228cbd47b8e888371052442acef36587e626634..85dae96ef29d927781b0cfe84ab7935cac785980) - Addressed review, changed commit messages, separated introduction of `newAdmissionReviewForVMIMCreation`.

[Change](https://github.com/kubevirt/kubevirt/compare/85dae96ef29d927781b0cfe84ab7935cac785980..7741afa91a5f9894e0f927e8c44c92c605af9542) - Remove no-longer relevant commit message.

orelmisan commented on pkg/virt-api/webhooks/validating-webhook/admitters/migration-create-admitter_test.go:
```diff
@@ -23,52 +23,34 @@ import (
-		mockVMIClient.EXPECT().Get(gomock.Any(), inFlightMigration.Spec.VMIName, gomock.Any()).Return(vmi, nil)
-		migrationInterface.EXPECT().List(gomock.Any(), gomock.Any()).Return(kubecli.NewMigrationList(inFlightMigration), nil).AnyTimes()
+		virtClient := kubevirtfake.NewSimpleClientset(vmi, inFlightMigration)
+		migrationCreateAdmitter := NewMigrationCreateAdmitter(virtClient)
```
Could you please move the admitter's instantiation closer to its usage?
Same bellow.

orelmisan commented on pkg/virt-api/webhooks/validating-webhook/admitters/migration-create-admitter_test.go:
```diff
@@ -188,17 +167,17 @@ var _ = Describe("Validating MigrationCreate Admitter", func() {
 			migration := v1.VirtualMachineInstanceMigration{
 				ObjectMeta: metav1.ObjectMeta{
-					Namespace: "default",
+					Namespace: vmi.Namespace,
```
IIRC, `libvmi.New()` returns a VMI with an unset Namespace.
There is an option `libvmi.WithNamespace`.

orelmisan commented on pkg/virt-api/webhooks/validating-webhook/admitters/migration-create-admitter_test.go:
```diff
@@ -268,3 +226,19 @@ var _ = Describe("Validating MigrationCreate Admitter", func() {
+		return nil, err
+	}
+
+	return &admissionv1.AdmissionReview{
```
nit: Please consider adding `Operation: "CREATE",`

orelmisan reviewed (commented):
Thank you for the changes.

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/7741afa91a5f9894e0f927e8c44c92c605af9542..97fd4fec4efa4aa118e374d1aec64603822eecb1) - Added an explanation for the addition of labels to the migration object, Added `libvmi.WithNamespace` option to create VMIs in the default namespace, Moved the migration admitter instantiation closer to its use, Added appropriate operation to the admission review result.

Acedus commented:
/test all

orelmisan reviewed (approved):
Thank you @Acedus

dhiller commented:
/cc @iholder101

iholder101 commented on pkg/virt-api/webhooks/validating-webhook/validating-webhook.go:
```diff
@@ -56,7 +56,7 @@ func ServeVMIPreset(resp http.ResponseWriter, req *http.Request) {
 
 func ServeMigrationCreate(resp http.ResponseWriter, req *http.Request, virtCli kubecli.KubevirtClient) {
-	validating_webhooks.Serve(resp, req, admitters.NewMigrationCreateAdmitter(virtCli))
+	validating_webhooks.Serve(resp, req, admitters.NewMigrationCreateAdmitter(virtCli.GeneratedKubeVirtClient()))
```
I see this is also done in https://github.com/kubevirt/kubevirt/pull/11897.

Is there a reason not to change `virtAPIApp`'s virtCli to the generated client in general, hence do it for all webhooks altogether?

orelmisan replied:
Since there are unit test using gomock, this is not a trivial change.
It requires change also to the unit tests.

iholder101 commented on pkg/virt-api/webhooks/validating-webhook/admitters/migration-create-admitter_test.go:
```diff
@@ -267,3 +223,20 @@ var _ = Describe("Validating MigrationCreate Admitter", func() {
 	})
 })
+
+func newAdmissionReviewForVMIMCreation(migration *v1.VirtualMachineInstanceMigration) (*admissionv1.AdmissionReview, error) {
+	migrationBytes, err := json.Marshal(migration)
+	if err != nil {
+		return nil, err
+	}
+
+	return &admissionv1.AdmissionReview{
+		Request: &admissionv1.AdmissionRequest{
+			Resource: webhooks.MigrationGroupVersionResource,
+			Object: runtime.RawExtension{
+				Raw: migrationBytes,
+			},
+			Operation: admissionv1.Create,
+		},
+	}, nil
+}
```
You've referenced a [similar function](https://github.com/kubevirt/kubevirt/blame/main/pkg/virt-api/webhooks/mutating-webhook/mutators/migration-create-mutator_test.go#L75) for migration-create-mutator_test.go in the commit message. You also state that this is a follow-up to the PR that have introduced this function.

Instead of introducing many more PRs that will do the same, perhaps we can avoid this code duplication and create a single function to serve all, that can get as a parameter the groupVersionKind of the relevant resource?

Acedus replied:
I agree that a generic function can be created for that purpose, I'll check what can be done.
@orelmisan WDYT?

Acedus replied:
I think that this is a fine optimization that can be considered, but the goal is to get rid of gomock and make use of the generated clientset. Perhaps we can consolidate that logic at a later time once we complete the directive we've opted for in the issue, WDYT @iholder101?

orelmisan replied:
@iholder101 Similar functions exist for the other webhooks' tests that were recently refactored:
https://github.com/kubevirt/kubevirt/blob/4fdbdbfe85e565862e6d6a13807e2e9d5de68dfd/pkg/virt-api/webhooks/mutating-webhook/mutators/clone-create-mutator_test.go#L155

https://github.com/kubevirt/kubevirt/blob/4fdbdbfe85e565862e6d6a13807e2e9d5de68dfd/pkg/virt-api/webhooks/mutating-webhook/mutators/migration-create-mutator_test.go#L75

https://github.com/kubevirt/kubevirt/blob/4fdbdbfe85e565862e6d6a13807e2e9d5de68dfd/pkg/virt-api/webhooks/validating-webhook/admitters/migration-update-admitter_test.go#L232

https://github.com/kubevirt/kubevirt/blob/4fdbdbfe85e565862e6d6a13807e2e9d5de68dfd/pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter_test.go#L498

The advantage of having these functions per object and operation is the fact that it is easy to understand what their return value will be.
I personally think that trying to generalize them into a single function may reduce code readability, but nevertheless I'm in favor of exploring this direction in a follow-up.

Acedus replied:
After looking into it, I believe the effort in trying to create a generic function that handles all admission review creation would be more detrimental to code quality rather than beneficial (that's without taking into account the relatively complex logic such function would require in its implementation). Those functions, despite having near identical logic, serve different purposes, thus they don't violate the DRY principle.

enp0s3 replied:
@Acedus @orelmisan Hey folks. I think that the concern is real. I think that if we keep this PR as is, we will add another function to the code that we would later need to refactor again. The current state of this PR is that it's easy to understand, and I don't think that if you will add a generic package to create admission review package it will make this PR complicated. Vice versa, we can stop that train of PRs with duplicated functions now and provide a solution here. I will review it

Acedus replied:
Hi @enp0s3, that's fine by me. I think it would better fit a separate PR so I'll do that.
My idea was something akin to [libvmi](https://github.com/kubevirt/kubevirt/tree/main/pkg/libvmi), so something along the lines of:
```go
package libadmission

import (
	"encoding/json"

	admissionv1 "k8s.io/api/admission/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime"
)

// Option represents an action that enables an option.
type Option func(ar *admissionv1.AdmissionReview)

// New instantiates a new AdmissionReview configuration,
// building its properties based on the specified With* options.
func NewAdmissionReview[T any](object *T, opts ...Option) (*admissionv1.AdmissionReview, error) {
	ar, err := baseAdmissionReview(object)
	if err != nil {
		return nil, err
	}

	for _, f := range opts {
		f(ar)
	}

	return ar, nil
}

func WithOperation(operation admissionv1.Operation) Option {
	return func(ar *admissionv1.AdmissionReview) {
		ar.Request.Operation = operation 
	}
}

func baseAdmissionReview[T any](object *T) (*admissionv1.AdmissionReview, error) {
	objectBytes, err := json.Marshal(object)
	if err != nil {
		return nil, err
	}

	return &admissionv1.AdmissionReview{
		Request: &admissionv1.AdmissionRequest{
			Object: runtime.RawExtension{
				Raw: objectBytes, 
			},
		},
	}, nil
}
...
```

WDYT?

enp0s3 replied:
@Acedus Hi, yes exactly! IMO it would be great to have this approach.

enp0s3 replied:
@Acedus Although I had a conversation with @orelmisan yesterday and seems that it could be difficult to make it generic, for example there is a need to mention the CRD name and version. But if you will make it work it could be great!

orelmisan replied:
@Acedus @enp0s3, this approach will not improve code readability IMO.
Imagine having a call with a complex builder, one could argue that it is too complex, wrap it in a function to simplify it - and then we are back to square one but with an unreadable function.

Is it possible to move forward with this PR and explore additional options in a follow-up?

enp0s3 replied:
@orelmisan SGTM

orelmisan replied:
```go
actualResponse := admitter.Admit(newAddmissionReview(WithOption1(), WithOption2(),WithOption3()))
```

One will probably want to warp it:
```go
func newAdmissionReviewForVMIMCreation(vmim *VirtualMachineInterfaceMigration) *admissionv1.AdmissionReview {
    return newAddmissionReview(WithOption1(), WithOption2(),WithOption3()
}
```

Acedus replied:
I agree with @orelmisan that it would simply result in more wrapping despite the possible extensibility of it, one gripe I had with it personally is that unlike libvmi it doesn't really have anything to do with KubeVirt. 

As for GVR/GVK, this is just a general question (perhaps I'm missing something) but can't we rely on the fact that all tested objects satisfy the `ObjectKind` interface and have the `GroupVersionKind` method?

In any event, I'm indifferent and wouldn't mind circling back to the implementation of a generic way to construct admission reviews later should we need it. Thanks for the thorough review!

iholder101 reviewed (commented):
Thanks a lot @Acedus!

I think this PR is very similar to https://github.com/kubevirt/kubevirt/pull/11897. My concern is that we'll have many follow-up PRs that look almost the same. While it's great to improve the code incrementally, there's also "cost" to it, e.g. in the form of review time.

See comment below about generally replace the virCli for example.
I also wonder if we can go for a single commit to unexport variables and introduce constructors for more admitters. While it would change more files, the change is trivial enough to be easily reviewed this way IMO.

orelmisan commented:
Hi @iholder101, is there anything else that should be done in order to move forward with this PR?

iholder101 commented:
> Hi @iholder101, is there anything else that should be done in order to move forward with this PR?

Hi @orelmisan!
I'm just overloaded at the moment, so will have to circle back to it when I have more bandwidth. I'll uncc myself for now, feel free to ping another approver in the meantime!

/uncc

enp0s3 commented:
/cc

enp0s3 commented:
/approve
